### DISCUSSION TITLE
Please limit the amount and scope of what your extension can do with the privilege it has.

### DISCUSSION THREAD

**Discussion by mudapanda2** [d8ahazard/sd_dreambooth_extension#1191] [Ideas]
As of now development on SD Webui seems to be on hold for whatever reason.

Most end users will not be aware of the scope an extension has when it comes to changing their installation as this isn't something one typically finds in an extension. Changing core dependencies, adding scripts that REQUIRE an internet connection, etc. There's no oversight and there's no control here in an arena that is normally at best a dependency circus. 

Out of the multitude of guides written out there none of them will tell end users that installing Dreambooth will fundamentally change what they've already been installing as per those guides. It ends up with unresolved broken dependencies and the common answer to questions of why things aren't working to be uninstalling this extension.

I have no doubt you work very hard at this and you've done very well but you're being overzealous and inconsiderate with the changes that will be made when somebody clicks a button believing they're installing a simple add-on, nobody is stopping you from forking sd webui so you can exercise that level of control. With seemingly nobody at the helm of the sd webui repo and the internet being full of guides, videos and otherwise pointing people at it some of the conflicts and major dependency changes this extension pushes should be rolled back.

**Comment by KickAssDave** [comment]
I have to agree with your comments.

I ended up in a right mess trying to get this to work alongside the requirements for the webui and a load of other dependencies.

I'm now at a point where things are working again, but yeah.... A few warnings would have helped.

**Comment by d8ahazard** [comment]
This is not a new discussion or set of observations. And, while I completely understand and agree with the sentiments, my retort has always been "Then maybe the extension system should have been designed better." 

As a developer, if I have the unfettered ability to override requirements and change core methods of the application for my benefit, then the logical thing would be to do that.

Or, as an additional argument, I'd point out that Auto1111 does this in the main application a LOT. As an example - the "Safe Unpickler" module *completely* overrides the default torch behavior across the board, which has caused a myriad of issues for me when developing this and other extensions.

Or for another example - xformers. If you don't explicitely specify the xformers flag (this may have changed recently), then even if you have xformers installed - extension code will not be able to load it. As in, Auto1111 completely and utterly breaks xformers behavior entirely if you don't specify this flag. It's crazy.


Now, this isn't meant as a cop out to the behavior designed into the extension. I'm sure there are places we could peel back some of the things that override some of the things...but if the underlying application allows these sort of things to happen, then I feel like some of the impetus should be on the app's developer(s) to ensure that extensions can't break the main app.

This, and a myriad of other grumbles with Auto1111 are the reason I've been focusing primarly on creating a new UI from the ground-up that is intended as a completely new web application for doing "all the things" with Stable-Diffusion...but in a security-focused and user-centric environment. 

Extensions are restricted to using specific handlers exposed to them and can't override default system requirements, there is no Gradio, and everything is designed with modularity in mind. If an extension needs to save an image or load a file, there are handlers for that. If it needs to communicate with the webUI, load a model, or run a GPU-intensive task - there are handlers for that.

While it's still very much a work-in-progress, at this point, I have dreambooth training working, I have a module for browsing and manipulating files (including uploads), I have a basic image tagging module, and I have an inference module with working controlnet support, and inpainting/img2img in the works.

![image](https://user-images.githubusercontent.com/1633844/234595085-c943bdca-e73f-4c91-9ca9-50abb597e5c8.png)


Point being, I very much understand your points, and I hope you can understand that as a developer, I probably have far more issue with the way Auto1111 works than you do. It and issues like this drive me insane. So much, that I'd literally rather reinvent the wheel than keep trying to design an extension for an extension system that - like many other things in the app - were added as an afterthought.

My sincerest apologies for your frustrations, and I hope that in the very near future, discussions like these will just disappear because we're all using an app that was designed from the ground-up to just work all the time...

**Reply by KickAssDave** [reply]
Hey - I appreciate your honest comments and genuinely appreciate your work.  I don't code in Python, but have done plenty of coding and am fully aware of the complexity, time and headaches it can give you.

Sorry if my own response came across as moany or entitled.

**Reply by ClashSAN** [reply]
@d8ahazard I'm guessing this request and your issues with webui are unsolvable.
Is this something we should be making very clear in the description of extensions in the extensions index? (e.g. please install with a fresh standalone webui project with these parameters: 1, 2, 3)

**Reply by hithereai** [reply]
Yeah I think it's fair to tell the users they will most probably run into problems if they try to install any other extenison with DreamBooth enabled. I have a \*dedicated\* webui just for DB. Sad but it's how it is.

**Reply by d8ahazard** [reply]
I can't honestly say what the total impact is of trying to run the DB extension on Auto1111 these days. It's now been several weeks since I've used Auto1111 for anything.

In looking at the code for the extension, I think it's somewhat safe to just install it with everything else these days, I don't think there are any drastic monkey-patches going on or crazy dependency overrides.

Honestly, IMHO, if you wanna run Dreambooth at this point, just use the new UI. It's not 100% done, but it's probably good enough for most people's training purposes. And it's not just for Dreambooth...I've added all kinds of bells and whistles.

**Reply by ClashSAN** [reply]
Thanks, I hope that you and your fellows get above the negative attitudes present in the repo, and enjoy working on the newest project.

Some devs: such as https://github.com/vladmandic/automatic, want to upgrade the extensions index. That project in particular has training available from the get-go, so this extension should be just for webui.

- ~~I think what will happen in the future is the extension when installed will auto downgrade webui to 1.1.0 via specified metadata.~~
- Do you prefer not to maintain the extension? The tutorials and guides across the internet point to the webui extension, and not to your newest project.
- What kind of expectation in communication should be had between webui extension developers and the project? (This is just proper feedback) I went in the webui dev discord group to say hello, and ask questions. If you can't get help on github, I'll handle it if you need.

**Reply by d8ahazard** [reply]
For now, we are still maintaining the extension, and will continue to until it becomes unreasonable to do so. It would definitely make my life easier to not have to live in a "split world", but as of right now, it's not too painful.

Not sure what you mean re: communication.
